### PR TITLE
Add payment history for probability calculation in path finding 

### PR DESCRIPTION
**Pull request by chenyukang**
#228 is left over because of an unnoticed merge.

### CODE REVIEW

File: src/fiber/network.rs [hunk 1]
```diff
@@ -2216,29 +2220,39 @@ where
             if payment_session.status == PaymentSessionStatus::Inflight {
                 match reason {
                     RemoveTlcReason::RemoveTlcFulfill(_) => {
+                        self.network_graph
+                            .write()
+                            .await
+                            .record_payment_success(&payment_session);
                         payment_session.set_success_status();
```
Comment (quake) [inline src/fiber/network.rs]: suggest to move these 2 lines to NetworkGraph#record_payment_success fn
```
payment_session.set_success_status();
self.store.insert_payment_session(payment_session);
```

File: src/fiber/network.rs [hunk 2]
```diff
@@ -2216,29 +2220,39 @@ where
...
-                        {
+                        let error_detail = reason.decode().expect("decoded error");
+                        self.update_graph_with_tlc_fail(&error_detail).await;
+                        let need_to_retry = self
+                            .network_graph
+                            .write()
+                            .await
+                            .record_payment_fail(&payment_session, error_detail.clone());
+                        if payment_session.can_retry() && need_to_retry {
                             let res = self.try_payment_session(state, payment_session).await;
```
Comment (quake) [inline src/fiber/network.rs:2237]: same suggestion as record_payment_success:

```
let need_to_retry = network_graph.record_payment_fail(...);
if need_to_retry {
  self.try_payment_session(...)
}
```

File: src/fiber/network.rs [hunk 3]
```diff
@@ -2285,73 +2302,142 @@ where
...
+            Ok(packet) => packet,
+            Err(e) => {
+                let err = format!(
+                    "Failed to create onion packet: {:?}, error: {:?}",
+                    payment_data.payment_hash, e
+                );
+                self.set_payment_fail_with_error(payment_session, &err);
+                return Err(Error::SendPaymentError(err));
+            }
+        };
```
Comment (quake) [inline src/fiber/network.rs]: suggest to move this block code after line 2361 which will avoid unnecessary clone on `hops`

File: src/fiber/network.rs [hunk 4]
```diff
@@ -2285,73 +2302,142 @@ where
...
+        let session_route =
+            SessionRoute::new(state.get_public_key(), payment_data.target_pubkey, &hops);
+
+        let (send, recv) = oneshot::channel::<Result<u64, TlcErrPacket>>();
+        let rpc_reply = RpcReplyPort::from(send);
+        let command = SendOnionPacketCommand {
+            packet: peeled_packet.serialize(),
+            previous_tlc: None,
+        };
+        payment_session.route = session_route.clone();
```
Comment (quake) [inline src/fiber/network.rs]: unnecessary clone

File: src/fiber/graph.rs [hunk 5]
```diff
@@ -748,31 +782,27 @@ where
...
-            } else {
-                break;
-            }
+        while let Some(elem) = distances.get(&current) {
+            let next_hop = elem.next_hop.as_ref().expect("next_hop is none");
+            result.push(PathEdge {
+                target: next_hop.0,
+                channel_outpoint: next_hop.1.clone(),
+            });
+            current = next_hop.0;
```
Comment (quake) [inline src/fiber/graph.rs]: suggest to use remove:
```suggestion
        while let Some(elem) = distances.remove(&current) {
            let (next_pubkey, next_out_point) = elem.next_hop.expect("next_hop is none");
            result.push(PathEdge {
                target: next_pubkey,
                channel_outpoint: next_out_point,
            });
            current = next_pubkey;
            if current == target {
                break;
            }
        }
```

File: src/fiber/graph.rs [hunk 6]
```diff
@@ -826,6 +859,60 @@ pub enum PaymentSessionStatus {
...
+        router
+    }
+
+    fn add_node(&mut self, pubkey: Pubkey, channel_outpoint: OutPoint, amount: u128) {
+        self.nodes.push(SessionRouteNode {
+            pubkey,
+            channel_outpoint,
+            amount,
+        });
+    }
```
Comment (quake) [inline src/fiber/graph.rs]: suggest to use zip to simplify code and remove unnecessary vec allocation code, btw, the type of payment_hops can be simplified to slice:

```suggestion
    pub fn new(source: Pubkey, target: Pubkey, payment_hops: &[PaymentHopData]) -> Self {
        let nodes = std::iter::once(source)
            .chain(payment_hops.iter().map(|hop| hop.next_hop.clone().unwrap_or(target)))
            .zip(payment_hops)
            .map(|(pubkey, hop)| SessionRouteNode {
                pubkey,
                channel_outpoint: hop.channel_outpoint.clone().unwrap_or_default(),
                amount: hop.amount,
            })
            .collect();

        Self { nodes }
    }
```

File: src/fiber/history.rs [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+    pub fn add_fail_pair(&mut self, from: Pubkey, target: Pubkey) {
+        self.add(from, target, current_time(), 0, false);
+        self.add(target, from, current_time(), 0, false)
+    }
+
+    pub fn add_fail_pair_balanced(&mut self, from: Pubkey, target: Pubkey, amount: u128) {
+        self.add(from, target, current_time(), amount, false);
+    }
+
+    pub fn fail_node(&mut self, nodes: &Vec<SessionRouteNode>, index: usize) {
```
Comment (quake) [inline src/fiber/history.rs]: all the type of `nodes` may be changed to slice.

File: src/fiber/history.rs [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+
+    #[cfg(test)]
+    pub(crate) fn reset(&mut self) {
+        self.inner.clear();
+    }
+
+    pub(crate) fn add_result(&mut self, from: Pubkey, target: Pubkey, result: TimedResult) {
+        self.inner
+            .entry(from)
+            .or_insert_with(HashMap::new)
```
Comment (quake) [inline src/fiber/history.rs]:
```suggestion
.or_default()
```

File: src/fiber/history.rs [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+    }
+
+    pub(crate) fn load_from_store(&mut self) {
+        let results = self.store.get_payment_history_result();
+        for (from, target, result) in results.iter() {
+            self.inner
+                .entry(from.clone())
+                .or_insert_with(HashMap::new)
+                .insert(target.clone(), *result);
+        }
```
Comment (quake) [inline src/fiber/history.rs]:
```suggestion
        for (from, target, result) in results.into_iter() {
            self.inner
                .entry(from)
                .or_default()
                .insert(target, result);
        }
```

File: src/fiber/history.rs [hunk 10]
```diff
@@ -0,0 +1,557 @@
...
+            }) {
+                if from == fail_node || target == fail_node {
+                    pairs.push((from, target));
+                }
+            }
+            for (from, target) in pairs {
+                self.apply_pair_result(from, target, 0, false, current_time());
+            }
+        }
+    }
```
Comment (quake) [inline src/fiber/history.rs]:
```suggestion
    pub(crate) fn apply_internal_result(&mut self, result: InternalResult) {
        let InternalResult {
            pairs,
            fail_node,
        } = result;

        for ((from, target), pair_result) in pairs.into_iter() {
            self.apply_pair_result(
                from,
                target,
                pair_result.amount,
                pair_result.success,
                pair_result.time,
            );
        }

        if let Some(fail_node) = fail_node {
            let pairs: Vec<Pubkey, Pubkey> = self.inner.iter().flat_map(|(from, targets)| {
                targets.keys().filter_map(move |target| (*from == fail_node || *target == fail_node).then_some((from.clone(), target.clone())))
            }).collect();

            for (from, target) in pairs {
                self.apply_pair_result(from, target, 0, false, current_time());
            }
        }
    }
```

**Comment by chenyukang** [comment]
addressed all issues, thanks @quake .

**Pull request merged**